### PR TITLE
Ports the player ticket panel from skyrat

### PR DESCRIPTION
**Pull request by SuperNovaa41**
## About The Pull Request

Ports the player ticket panel from Skyrat (https://github.com/Skyrat-SS13/Skyrat-tg/pull/2958)

https://user-images.githubusercontent.com/16159590/168328901-feb250f3-85e2-477a-998a-d826c0035c34.mp4

## Why It's Good For The Game
The ticket panel is a very useful tool for admins, as it allows you to view the entire ticket history.
Meanwhile, players can only view whatever pops up in chat.
Especially with TGchat not keeping the entire chat history, this can make it difficult (and time consuming) for players to search back through chat to see what admins have said to them, and what they themselves have said.
Giving players an easy to access panel that shows them their whole ticket history is very useful, I think.

## Changelog

:cl: NotRanged, SuperNovaa41
qol: There is a new verb in the Admin tab that lets you view your open Admin Help ticket.
/:cl:

### CODE REVIEW

**Comment by SmArtKar** [comment]
Now TGUI it.

**Comment by wesoda25** [comment]
Would tgui slow it down at all? Ticky panel already pretty slow I've found, unless this functions differently.

**Comment by optimumtact** [comment]
have you informed skyrat that you're upstreaming this, sinc ethey will have to remove their version

**Comment by Pepsilawn** [comment]
With this PR the user is made aware whether the ticket is ongoing or Closed, before this PR the ticketed user wouldn't be aware when it's marked as such as opposed to choosing to Resolve or Reject which notify them in chat. I'm not sure if the user should have access to that information as that button only seem to serve to close the ticket without the other person knowing.

**Comment by Iamgoofball** [comment]
players have the right to know if their ticket is closed

**Comment by SuperNovaa41** [comment]
> With this PR the user is made aware whether the ticket is ongoing or Closed, before this PR the ticketed user wouldn't be aware when it's marked as such as opposed to choosing to Resolve or Reject which notify them in chat. I'm not sure if the user should have access to that information as that button only seem to serve to close the ticket without the other person knowing.

Talked to a few admins about this. They've told me they only use it in cases where the tickets are old and were forgotten to be closed. If this IS an issue I'm not sure how I'd go about addressing it.

**Comment by SuperNovaa41** [comment]
> have you informed skyrat that you're upstreaming this, sinc ethey will have to remove their version

No, but i will

File: code/modules/admin/verbs/adminhelp.dm [hunk 1]
```diff
@@ -509,6 +517,7 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 	message_admins(msg)
 	log_admin_private(msg)
 	AddInteraction("Rejected by [key_name].")
+	add_player_interaction("Rejected by [key_name_admin(usr, FALSE)].")
```
Comment (LemonInTheDark) [inline code/modules/admin/verbs/adminhelp.dm:520]: This "by" behavior is new right? players didn't used to see that. (Same can be said for most of this)

File: code/modules/admin/verbs/adminhelp.dm [hunk 2]
```diff
@@ -528,6 +537,7 @@ GLOBAL_DATUM_INIT(ahelp_tickets, /datum/admin_help_tickets, new)
 	message_admins(msg)
 	log_admin_private(msg)
 	AddInteraction("Marked as IC issue by [key_name]")
+	add_player_interaction("Marked as IC issue by [key_name_admin(usr, FALSE)]")
```
Comment (LemonInTheDark) [inline code/modules/admin/verbs/adminhelp.dm:540]: How much overlap does `AddInteraction` have with `add_player_interaction`?

**Comment by JohnFulpWillard** [comment]
> players have the right to know if their ticket is closed

That's what the Resolve button is for. Closing is supposed to be silent, so if you plan on adding this PR then you might as well remove the Close ticket button entirely.

Also this says the admin's IC name which I am not a fan of.

**Comment by SuperNovaa41** [comment]
> Why is there no refresh button/why does sending a message not refresh the window?

Not refreshing when a new message is received is consistent with the current admin ticket panel. I'll add a refresh button

File: code/modules/admin/verbs/adminhelp.dm (hunk 1 above)
Comment (SuperNovaa41) [reply code/modules/admin/verbs/adminhelp.dm:520]: Yeah, i'll change up these to be more accurate to what it was. This was just a direct port tbh

**Comment by SuperNovaa41** [comment]
Okay fixed a few things, no longer says admin IC names, and there is a functional refresh button

**Comment by SuperNovaa41** [comment]
> That's what the Resolve button is for. Closing is supposed to be silent, so if you plan on adding this PR then you might as well remove the Close ticket button entirely.

image

aswell, i don't think theres a reason that we should be tricking players into thinking their ticket is still open when, it isn't.

this still won't notify players that it's been closed, they'll just be able to see that it was if they take the time to look at the panel.

File: code/modules/admin/verbs/adminhelp.dm [hunk 3]
```diff
@@ -761,22 +819,28 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
 //what can be a client, ckey, or mob
 //log_in_blackbox: Whether or not this message with the blackbox system.
 //If disabled, this message should be logged with a different proc call
-/proc/admin_ticket_log(what, message, log_in_blackbox = TRUE)
+/proc/admin_ticket_log(what, message, log_in_blackbox = TRUE, player_message)
```
Comment (tralezab) [inline code/modules/admin/verbs/adminhelp.dm:822]: doc this new arg! also to consider: should it be after `log_in_blackbox`?

File: code/modules/admin/verbs/adminhelp.dm [hunk 4]
```diff
@@ -761,22 +819,28 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
...
 	var/client/C
 	var/mob/Mob = what
 	if(istype(Mob))
 		C = Mob.client
 	else
 		C = what
 	if(istype(C) && C.current_ticket)
-		C.current_ticket.AddInteraction(message)
+		if (isnull(player_message))
+			C.current_ticket.AddInteraction(message)
```
Comment (tralezab) [inline code/modules/admin/verbs/adminhelp.dm:831]: Better var name than `C` please since ya use it

File: code/modules/admin/verbs/adminhelp.dm [hunk 5]
```diff
@@ -761,22 +819,28 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
...
+			C.current_ticket.AddInteraction(message, player_message)
 		if(log_in_blackbox)
 			SSblackbox.LogAhelp(C.current_ticket.id, "Interaction", message, C.ckey, usr.ckey)
 		return C.current_ticket
 	if(istext(what)) //ckey
 		var/datum/admin_help/AH = GLOB.ahelp_tickets.CKey2ActiveTicket(what)
 		if(AH)
-			AH.AddInteraction(message)
+			if (isnull(player_message))
+				AH.AddInteraction(message)
```
Comment (tralezab) [inline code/modules/admin/verbs/adminhelp.dm:841]: Better var name than `AH` please since ya use it

File: code/modules/admin/verbs/adminpm.dm [hunk 6]
```diff
@@ -214,7 +214,7 @@
 			type = MESSAGE_TYPE_ADMINPM,
 			html = span_notice("PM to-<b>Admins</b>: <span class='linkify'>[rawmsg]</span>"),
 			confidential = TRUE)
-		var/datum/admin_help/AH = admin_ticket_log(src, "<font color='red'>Reply PM from-<b>[key_name(src, TRUE, TRUE)]</b> to <i>External</i>: [keywordparsedmsg]</font>")
+		var/datum/admin_help/AH = admin_ticket_log(src, "<font color='red'>Reply PM from-<b>[key_name(src, TRUE, TRUE)]</b> to <i>External</i>: [keywordparsedmsg]</font>", player_message = "<font color='red'>Reply PM from-<b>[key_name(src, TRUE, FALSE)]</b> to <i>External</i>: [msg]</font>")
```
Comment (tralezab) [inline code/modules/admin/verbs/adminpm.dm:217]: Better var name than `AH` please since ya use it

File: code/modules/admin/verbs/adminhelp.dm [hunk 7]
```diff
@@ -752,6 +783,33 @@ GLOBAL_DATUM_INIT(admin_help_ui_handler, /datum/admin_help_ui_handler, new)
 	GLOB.admin_help_ui_handler.ui_interact(mob)
 	to_chat(src, span_boldnotice("Adminhelp failing to open or work? <a href='?src=[REF(src)];tguiless_adminhelp=1'>Click here</a>"))
 
+/client/verb/viewlatestticket()
```
Comment (tralezab) [inline code/modules/admin/verbs/adminhelp.dm:786]: snake case for this verb, lad!

**Review by tralezab — CHANGES_REQUESTED** [review]
oh so minor things

File: code/modules/admin/verbs/adminhelp.dm (hunk 3 above)
Comment (SuperNovaa41) [reply code/modules/admin/verbs/adminhelp.dm:822]: oops, probably not

**Comment by LemonInTheDark** [comment]
You should credit the initial author in the changelog

**Pull request merged**